runtime/sema/authaccount.cdc

```diff
@@ -106,6 +106,14 @@ pub struct AuthAccount {
     /// The path must be a storage path, i.e., only the domain `storage` is allowed
     pub fun borrow<T: &Any>(from: StoragePath): T?
 
+    /// Returns true if the object in account storage under the given path satisfies the given type, 
+    /// i.e. could be borrowed using the given type.
+    ///
+    /// The given type must not necessarily be exactly the same as the type of the borrowed object.
+    ///
+    /// The path must be a storage path, i.e., only the domain `storage` is allowed.
+    pub fun check<T: &Any>(from: StoragePath): Bool
```

Comment:
The API here sounds a bit confusing to me. The reason is, references cannot be stored in a storage path, so checking whether the value stored belongs to a reference type doesn't sounds much sense. 

I think either the type parameter must be a non-reference type (i.e: `check<T: Any>` and the implementation should check whether the value belongs to concrete type `T`), OR we should maybe change the function name/description, to indicate that this API tries to check whether a reference can be borrowed from the given type `T` for the given path.

Reply:
> /// The given type must not necessarily be exactly the same as the type of the borrowed object.

My concern here is that it still is ambiguous: it says the type doesn't have to be the exact type, but doesn't say what the relationship between type `T` and the value stored at the path.

Just to add more: all other storage-related functions such as `load`, `copy`, and `save` functions (except `borrow`) take a non-reference type `T`. Similarly, I feel it's more-natural/less-ambiguous to have a non-reference type as the type-parameter. Such that, `check<T>(/some/path)` checks whether the value at the path is of type `T`. 
i.e: I feel this function should be the inverse check of `AuthAccount.typeAt()` (https://github.com/onflow/cadence/pull/1254)

Reply:
As far as I understood @joshuahannan's request in #564, the function is meant to determine if the path could be `borrow`ed with the given type. Maybe the main issue here is that the `check` name is potentially confusing.

Reply:
Yeah, either one of `canBorrow<T: &Any>()` or `check<T: Any>()` is OK to me. I think both of these should facilitate the requirement of https://github.com/onflow/cadence/issues/564.
Although I have a slight preference for `check` purely on naming-elegancy. But fine with either.

Reply:
@joshuahannan What are your thoughts?

Reply:
I like `check` better also since it aligns with the `capability.check`. Also, will this work for stored capabilities, or will it only work for stored resources and structs?

like if I have a `@FlowToken.Vault` in a storage path, I can call `check` on the path and it'll return true, but if I have a valid `Capability<FlowToken.Vault>` in a storage path and I try to `check` it, will it also return true? That would be nice also

Reply:
@joshuahannan No, there is no such automatic/magic behaviour of handling stored capabilities specially.

@SupunS  `check<T>(path)` is basically then just sugar for `AuthAccount.type(at: path).isSubtype(of: Type<T>())`?

Reply:
yes, that's what I was thinking

Reply:
Yeah, will change the signature to `check<T>(path)` for this PR. @joshuahannan I think a new issue can be filed for that.

Reply:
Resolving this conversation.